### DISCUSSION TITLE
Question about MetaLayer and directed edges

### DISCUSSION THREAD

**Discussion by varghele** [pyg-team/pytorch_geometric#8262] [Q&A] [answered]
I have a question about the current MetaLayer implementation, especially the node model. In short: does the NodeModel of the MetaLayer take directional edges into account?

In long:
 Currently, I am trying to build a model that works on a directed graph, where information can only be passed "left to right" so to speak.
When creating a graph, it is advised to build two edges with the edge indices e.g. 0->1 and 1->0, if one wants a bidirectional edge. Ok no problem, I want a unidirectional one, so I just leave out one of the edge index pairs. I can even leave out node information on the edge (by not including source or target in the EdgeModel class).

However now I have an issue with understanding the code of the NodeModel (see below, taken from docu of 2.4.0). To my understanding it aggregates over ALL edges connected to the node regardless of their directionality, while I would want it to only take into account edges pointing towards it.
Am I misunderstanding something with row before the scatter and col in the scatter? It very much feels like I am getting something wrong. Thanks for the help :)

```python
class NodeModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.node_mlp_1 = Seq(Lin(..., ...), ReLU(), Lin(..., ...))
        self.node_mlp_2 = Seq(Lin(..., ...), ReLU(), Lin(..., ...))

    def forward(self, x, edge_index, edge_attr, u, batch):
        # x: [N, F_x], where N is the number of nodes.
        # edge_index: [2, E] with max entry N - 1.
        # edge_attr: [E, F_e]
        # u: [B, F_u]
        # batch: [N] with max entry B - 1.
        row, col = edge_index
        out = torch.cat([x[row], edge_attr], dim=1)
        out = self.node_mlp_1(out)
        out = scatter(out, col, dim=0, dim_size=x.size(0),
                      reduce='mean')
        out = torch.cat([x, out, u[batch]], dim=1)
        return self.node_mlp_2(out)
```

**Comment by rusty1s** [comment, marked as answer]
All graphs in PyG are understood as directional. For bidirectional edges, one has to add both directions to the `edge_index`. As such, PyG already does what you want it to do.

**Reply by varghele** [reply]
Thanks, that helps :)
